File: src/_02_StructuralPattern/_02_02_Bridge/DefaultChampion.js
import Champion from "./Champion.js";

/**
 * @typedef { import("./Skin").default } Skin
 */

export default class DefaultChampion extends Champion {
  /**
   * 챔피언명
   * @type { string }
   */
  #name

  /**
   * @type { Skin }
   */
  #skin

  /**
   * @param { string } name
   * @param { Skin } skin
   */
  constructor(name, skin) {
    super();
    this.#name = name;
    this.#skin = skin;
  }

  /** @override */
  move() {
    console.log(`${this.#skin.getName()} ${this.#name} - move`);
  }

  /** @override */
  skillQ() {
    console.log(`${this.#skin.getName()} ${this.#name} Skill Q`);
  }
  
  /** @override */
  skillW() {
    console.log(`${this.#skin.getName()} ${this.#name} Skill W`);
  }

  /** @override */
  skillE() {
    console.log(`${this.#skin.getName()} ${this.#name} Skill E`);
  }
  
  /** @override */
  skillR() {
    console.log(`${this.#skin.getName()} ${this.#name} Skill R`);
  }
}

File: src/_02_StructuralPattern/_02_02_Bridge/java/after/DefaultChampion.java
package _02_StructuralPattern._02_02_Bridge.java.after;

public class DefaultChampion implements Champion {

  private String name;
  private Skin skin;

  public DefaultChampion(String name, Skin skin) {
    this.name = name;
    this.skin = skin;
  }
  
  @Override
  public void move() {
    System.out.printf("%s %s move\n", skin.getName(), name);
  }

  @Override
  public void skillQ() {
    System.out.printf("%s %s skillQ\n", skin.getName(), name);
  }

  @Override
  public void skillW() {
    System.out.printf("%s %s skillW\n", skin.getName(), name);
  }

  @Override
  public void skillE() {
    System.out.printf("%s %s skillE\n", skin.getName(), name);
  }

  @Override
  public void skillR() {
    System.out.printf("%s %s skillR\n", skin.getName(), name);
  }
  
}
